KmdCfg config validation and KmdCfg config tests

amityadav0 opened the pull request.

robert-zaremba commented on packages/algob/test/internal/core/config/config-validation.ts:
```diff
@@ -608,4 +608,148 @@ describe("Config validation", function () {
+      );
+    });
+
+    it("Shouldn't fail with valid KMD config", function () {
```
```suggestion
    it("Should work with valid KMD config", function () {
```

robert-zaremba replied:
Let's don't use a double negations.

amityadav0 replied:
ok

robert-zaremba commented on packages/algob/test/internal/core/config/config-validation.ts:
```diff
@@ -608,4 +608,148 @@ describe("Config validation", function () {
+      assert.isEmpty(errors.errors, errors.toString());
+    });
+
+    it("Shouldn't fail with unrecognized params", function () {
```
```suggestion
    it("Should work with unrecognized params", function () {
```

robert-zaremba commented on packages/algob/test/internal/core/config/config-validation.ts:
```diff
@@ -608,4 +608,148 @@ describe("Config validation", function () {
+              }
+            ],
+            otherParam: "some_other_detail"
+          },
```
you can store a valid config in variable and reuse it in 2 tests instead of copying lot of lines.
Here, in this test you can do:
```
cfg.networks.localhost.kmdCfg.otherParam = "some_other_detail"
```

robert-zaremba replied:
In other tests you can copy the object using `deepmerge` (we already have it in dependencies):
```
let invalidCfg = deepmerge({}, cfg)
```

amityadav0 replied:
ok

amityadav0 replied:
![asd](https://user-images.githubusercontent.com/26249903/98079318-93986600-1e99-11eb-89dd-6598e0b8a0bc.png)

If I do it using `deepmerge`, it doesn't let me assign wrong types to values, but user can put wrong types in `algob.config.js`

robert-zaremba replied:
Ah, so maybe using a function will be a good solution?

```
let cfg = mkCfg()
```

amityadav0 replied:
![work](https://user-images.githubusercontent.com/26249903/98096655-06154000-1eb2-11eb-90db-ed992fe81ecc.png)

this works if i use any in tests.

amityadav0 replied:
![works](https://user-images.githubusercontent.com/26249903/98096879-3f4db000-1eb2-11eb-8be2-24b2252d670a.png)

this works if i use object.assign()

amityadav0 replied:
![work](https://user-images.githubusercontent.com/26249903/98096979-5e4c4200-1eb2-11eb-9bee-c559182967db.png)

this works if i use `any` with `object.assign()`

amityadav0 replied:
![work1](https://user-images.githubusercontent.com/26249903/98101463-411a7200-1eb8-11eb-8356-d067aa8089e3.png)

works when i use `any`

robert-zaremba reviewed (approved):
The configs in tests looks a bit cryptic. I though that `deepmerge` would make it easy and reduce the code to create a cfg to few lines. Feel free to finalize it as you like

amityadav0 commented:
understood, i have added a valid config at top of the tests and i am resusing this config using, `deepmerge` and `any` (as algob.config.js is in javascript). updated in https://github.com/scale-it/algorand-builder/pull/117/commits/30d11bac2a13bc1f8ccb549cebebf815243fc154 this commmit.
